File: starter-packs/python/network/decoder.py
from dataclasses import dataclass
from typing import Tuple
import struct
import uuid

from core.game_state import PlayerInfo, PlayerWeapon, Projectile, Blade, GameState, Coin
from core.map_state import Point, Collider, ColliderType, MapState


def read_str(byte_array, end_index=None):
    if end_index is None:
        end_index = byte_array.find(b'\0')
        if end_index == -1:
            return None 
    
    string = byte_array[:end_index].decode('utf-8')
    return string, end_index


def read_uuid(byte_array, end_index):
    return str(uuid.UUID(bytes=byte_array[:end_index]))


@dataclass
class JDISDecoder:
    def decode_point(self, data: bytes, offset: int) -> Tuple[Point, int]:
        p = Point()
        p.x, p.y = struct.unpack_from('<dd', data[offset:], 0)
        return p, offset + 16


    def decode_colliders(self, pos_size: int, data: bytes, offset: int) -> Tuple[Collider, int]:
        c = Collider()
        
        for i in range(pos_size):
            p = Point()
            p, _ = self.decode_point(data, i * 16)
            c.positions.append(p)

        offset += pos_size * 16
        c.collider_type = ColliderType(struct.unpack_from('<B', data, offset)[0])
        offset += 1
        
        return c, offset
    

    def decode_map_state(self, data: bytes) -> MapState:
        m = MapState()
        m.size = struct.unpack_from('<B', data, 0)[0]
        m.discrete_grid = []
        m.spawns = []
        m.walls = []

        # decode discrete grid
        m.discrete_grid = [
            list(struct.unpack_from('<' + 'B' * m.size, data, i * m.size + 1)) for i in range(m.size)
        ]

        # decode walls
        offset = m.size * m.size + 1
        walls_len = struct.unpack_from('<i', data, offset)[0]
        
        offset += 4
        for _ in range(walls_len):
            pos_size = struct.unpack_from('<B', data, offset)[0]
            offset += 1

            collider, offset = self.decode_colliders(pos_size, data, offset)

            m.walls.append(collider)

        m.save = bytearray(data[offset: offset + 100])

        return m


    def decode_player_info(self, data: bytes) -> Tuple[PlayerInfo, int]:
        p = PlayerInfo()

        p.name, end_index = read_str(data)
        offset = end_index + 1

        p.color, p.health, p.score = struct.unpack_from('<iiq', data, offset)
        offset += 16

        p.pos, offset = self.decode_point(data, offset)
        
        has_dest = struct.unpack_from('<?', data, offset)[0]
        offset += 1

        if has_dest:
            p.dest, offset = self.decode_point(data, offset)

        p.playerWeapon = PlayerWeapon(struct.unpack_from('<B', data, offset)[0])
        offset += 1

        projectile_size = struct.unpack_from('<i', data, offset)[0]
        offset += 4

        p.projectiles = []
        for _ in range(projectile_size):
            projectile = Projectile()
            projectile.uid = read_uuid(data[offset:], 16)
            offset += 16

            projectile.pos, offset = self.decode_point(data, offset)
            projectile.dest, offset = self.decode_point(data, offset)

            p.projectiles.append(projectile)

        p.blade = Blade()

        p.blade.start, offset = self.decode_point(data, offset)
        p.blade.end, offset = self.decode_point(data, offset)

        p.blade.rotation = struct.unpack_from('<d', data, offset)[0]
        offset += 8

        return p, offset
    

    def decode_game_state(self, data: bytes) -> Tuple[GameState, int]:
        g = GameState()
        g.current_tick, g.current_round = struct.unpack_from('<ib', data, 0)
        offset = 5

        player_size = struct.unpack_from('<i', data, offset)[0]
        offset += 4
        
        g.players = []
        for _ in range(player_size):
            player, n_offset = self.decode_player_info(data[offset:])
            offset += n_offset

            g.players.append(player)

        coin_size = struct.unpack_from('<i', data, offset)[0]
        offset += 4

        g.coins = []
        for _ in range(coin_size):
            coin = Coin()
            coin.uid = read_uuid(data[offset:], 16)
            offset += 16

            coin.pos, offset = self.decode_point(data, offset)

            coin.value = struct.unpack_from('<i', data, offset)[0]
            offset += 4

            g.coins.append(coin)

        return g

File: starter-packs/javascript/network/decoder.js
import { Buffer } from 'buffer';
import { PlayerInfo, PlayerWeapon, Projectile, Blade, GameState, Coin } from '../core/game_state.js';
import { Point, Collider, CollderType, MapState } from '../core/map_state.js';


const read_str = (buf, offset) => {
  const idx = buf.indexOf(0, offset);
  if (idx === -1)
    return [null, offset];
  const str = buf.toString('utf8', offset, idx);
  return [str, idx + 1];
};

const read_uuid = (buf, offset) => {
  const uuid = buf.slice(offset, offset + 16);
  return uuid.toString('hex').match(/.{1,8}/g).join('-');
};

class JDISDecoder {
  decode_point(data, offset) {
    const x = data.readBoudleLE(offset);
    const y = data.readBoudleLE(offset + 8);
    return [{ x, y }, offset + 16];
  }

  decode_colliders(pos_size, data, offset) {
    c = new Collider();

    for (let i = 0; i < pos_size; i++) {
      const [pos, new_offset] = this.decode_point(data, offset);
      c.points.push(pos);
      offset = new_offset;
    }

    c.collider_type = CollderType[data.readUInt8(offset)];
    offset += 1;

    return [c, offset];
  }

  decode_map_state(data) {
    const m = new MapState();
    m.size = data.readUInt8(0);
    m.discrete_grid = Array.from({ length: size }, (_, i) => 
      Array.from(data.slice(1 + i * size, 1 + (i + 1) * size))
    );

    let offset = 1 + size * size;
    const walls_len = data.readInt32LE(offset);
    offset += 4;

    for (let i = 0; i < walls_len; i++) {
      const pos_size = data.readUInt8(offset);
      offset += 1;

      const [c, new_offset] = this.decode_colliders(pos_size, data, offset);
      m.walls.push(c);
      offset = new_offset;
    }

    m.save = data.slice(offset, offset + 100);
    return m;
  }

  decode_player_info(data) {
    const p = new PlayerInfo();
    const [name, offset] = read_str(data, 0);
    p.name = name;

    p.color = data.readUInt32LE(offset);
    p.health = data.readFloatLE(offset + 4);
    p.score = data.readFloatLE(offset + 8);
    offset += 16;

    [p.pos, offset] = this.decode_point(data, offset);

    const has_dest = data.readUInt8(offset) !== 0;
    offset += 1;

    if (has_dest) {
      const [dest, new_offset] = this.decode_point(data, offset);
      p.dest = dest;
      offset = new_offset;
    }

    p.player_weapon = PlayerWeapon[data.readUInt8(offset)];
    offset += 1;

    const projectiles_len = data.readInt32LE(offset);
    offset += 4;

    for (let i = 0; i < projectiles_len; i++) {
      const projectile = new Projectile();
      [projectile.pos, offset] = this.decode_point(data, offset);
      [projectile.dest, offset] = this.decode_point(data, offset);
      p.projectiles.push(projectile);
    }

    [p.blade.start, offset] = this.decode_point(data, offset);
    [p.blade.end, offset] = this.decode_point(data, offset);

    return [p, offset];
  }

  decode_game_state(data) {
    const g = new GameState();

    g.current_tick = data.readInt32LE(0);
    g.current_round = data.readInt8(4);
    let offset = 5;

    const player_len = data.readInt32LE(offset);
    offset += 4;

    for (let i = 0; i < player_len; i++) {
      const [player, new_offset] = this.decode_player_info(data, offset);
      g.players.push(player);
      offset = new_offset;
    }

    const coins_len = data.readInt32LE(offset);
    for (let i = 0; i < coins_len; i++) {
      const coin = new Coin();
      coin.uid = read_uuid(data, offset);
      offset += 16;

      [coin.pos, offset] = this.decode_point(data, offset);
      coin.value = data.readInt32LE(offset);
      offset += 4;

      g.coins.push(coin);
    }

    return g;
  }
};

export { JDISDecoder };
